packages/controllers/src/multichain/MultiChainController.ts

```diff
@@ -0,0 +1,453 @@
+      ),
+    );
+  }
+
+  async onConnect(
+    origin: string,
+    connection: ConnectArguments,
+  ): Promise<Session> {
+    const existingSession = this.getSession(origin);
+    if (existingSession) {
+      await this.closeSession(origin);
+    }
+
+    const snaps = await this.messagingSystem.call('SnapController:getAll');
+    const filteredSnaps = snaps.filter((snap) => snap.enabled && !snap.blocked);
+
+    const availableNamespaces = await filteredSnaps.reduce<
+      Promise<Record<SnapId, Record<NamespaceId, Namespace>>>
+    >(async (previousPromise, snap) => {
+      const acc = await previousPromise;
+      const snapNamespaces = await this.snapToNamespaces(snap);
+      if (snapNamespaces !== null) {
+        return { ...acc, [snap.id]: snapNamespaces };
+      }
+      return acc;
+    }, Promise.resolve({}));
+
+    // The magical matching algorithm specified in SIP-2
+    const namespaceToSnaps = findMatchingKeyringSnaps(
+      connection.requiredNamespaces,
+      availableNamespaces,
+    );
+
+    const permissions = await this.messagingSystem.call(
+      'PermissionController:getPermissions',
+      origin,
+    );
+
+    const approvedNamespacesAndSnaps = Object.entries(namespaceToSnaps).reduce(
+      (acc, [namespace, snapIds]) => {
+        // If snap already is approved for use, solve conflict by using that snap
+        const approvedSnap = snapIds.find((snapId) => {
+          return (
+            permissions &&
+            hasProperty(permissions, getSnapPermissionName(snapId))
+          );
+        });
+
+        if (approvedSnap) {
+          return { ...acc, [namespace]: [approvedSnap] };
+        }
+        return acc;
+      },
+      {},
+    );
+
+    const hasConflicts = Object.keys(namespaceToSnaps).some(
+      (namespace) => !hasProperty(approvedNamespacesAndSnaps, namespace),
+    );
+
+    const filteredNamespacesAndSnaps = hasConflicts
+      ? namespaceToSnaps
+      : approvedNamespacesAndSnaps;
+
+    const possibleAccounts = await this.namespacesToAccounts(
+      origin,
+      filteredNamespacesAndSnaps,
+      connection.requiredNamespaces,
+    );
+
+    // If current configuration doesn't solve request, we need to show a prompt.
+    const resolvedAccounts = hasConflicts
+      ? await this.resolveConflicts(origin, possibleAccounts)
+      : possibleAccounts;
+    const chosenAccounts = fromEntries(
+      Object.entries(resolvedAccounts).map(([namespace, snapIds]) => [
+        namespace,
+        snapIds[0] ?? null,
+      ]),
+    );
+
+    const approvedPermissions = Object.values(chosenAccounts).reduce(
+      (acc, cur) => {
+        if (cur) {
+          return {
+            ...acc,
+            [getSnapPermissionName(cur.snapId)]: {},
+          };
+        }
+        return acc;
+      },
+      {},
+    );
+    // TODO: Save the approved permissions
+    await this.messagingSystem.call('PermissionController:grantPermissions', {
+      approvedPermissions,
```

Comment:
We shouldn't approve `wallet_invoke` just yet.

`onConnect` should have a new permission something like `origin_readableAccounts` that is a read-only access to accounts the user approved.

`snap_invoke` check should happen in `onRequest()`

Reply:
So you want an account selection prompt `onConnect` and a separate prompt to start doing requests?

Reply:
Do we have a design for this flow?

Reply:
Should `onRequest` trigger a prompt for connection to all requested snaps or just the one snap handling the request?

Reply:
No user facing UI designs for now, let's skip for Bogota unless Hassan and Nilsson have some free time before the end. But having the permission already will be very helpful later on.

Reply:
As discussed yesterday we'll use 1 permission for now. Leaving comments to refactor this for v2.

Reply:
Can you leave comments in the code?